Zng beta (#1394)
brimsec/zq: zng/docs/spec.md

> * the `bytes` type is not yet implemented,
> * the `enum` type is not yet implemented,
> * only streams of `record` types (which may consist of any combination of
>   other implemented types) are supported by zq even though a stream of
>   any types may currently be expressed in value messages.
>
> TBD: should we add maps? It would be a complex type comprising
> a list of key/values where the key type and value type is defined in the

> > map typedef.  The contrarian view is that maps are easily represented

Reviewer: Note that the contrarian view does not preclude having maps in ZQL at some point in the future. There could be a operator `map(list)` that turns a list of `record[key:keyType,value:valueType]`  into a map that can then be operated in by map-savvy procs and reducers in ZQL. Or for more flexibility, maybe `map()` takes the key/value field names, so you could say `map(list, key, value)`. Maps would have to be converted back into a list representation before exiting the flowgraph. 

There might also be a way to create maps from a stream of a records, for example `count() by _path | mapStream(_path, count)` would transform the stream of records into a histogram-like map of `string->int` which would be more suitable e.g. for statistical operations than separate records.

Anyway, spitballing here and not suggesting we need to do this... but it shows how the space of co-designing a data format and data language is novel broad. We're just scratching the outer surface here. 😎

> > TBD: should we add maps? It would be a complex type comprising
> > a list of key/values where the key type and value type is defined in the
> > map typedef.  The contrarian view is that maps are easily represented
> > as a sequence of record[key:keyType,value:valueType].

Reviewer: ```suggestion
```

a particular data value is specified by its "type identifier", or type ID,
which is an integer representing either a "primitive type" or a
"complex type".

The ZNG type system comprises the standard set of primitive types like integers,
floating point, strings, byte arrays, etc. as well as complex types
that are built from other types like records, arrays, and sets arranged
from the primitive types.

ZNG also includes first-class types, where a value can be of type "type".
This is useful for grouping aggregations by type, e.g., to do introspective data exploration.

Any value in ZNG can take on a null representation.  It is up to an implementation to
decide how to external data structures map into and out of values with nulls.  Typically,
a null value in a field column means an optional value was not present, though
the notion of optionality is not explicitly defined by ZNG.

Null values can be typed or untyped.  For example, a null value in a field of
a record has the type of the field's type.  A null can be "untyped" by creating
a null value of type null.  The only valid value of type null is null.

### Data Model Examples

> Here are a few examples of the ZNG data model expressed iin the TZNG text format.

Reviewer 2: ```suggestion
Here are a few examples of the ZNG data model expressed in the TZNG format.
```

that provide type definitions and other metadata.  The type of
a particular data value is specified by its "type identifier", or type ID,
which is an integer representing either a "primitive type" or a
"complex type".

The ZNG type system comprises the standard set of primitive types like integers,
floating point, strings, byte arrays, etc. as well as complex types
that are built from other types like records, arrays, and sets arranged
from the primitive types.

ZNG also includes first-class types, where a value can be of type "type".

> This is useful for introspection and grouping aggregations by type.

Reviewer: ```suggestion
This is useful for data exploration and grouping aggregations by type.
```

(maybe?)

Author: This reads funny to me when I just swap the terms.  I landed here
```
This is useful for grouping aggregations by type to do introspective data exploration.
```

with the following structure:
```
----------------------------------------------------------

> |0xf0|<nfields>|<field1><type-id-1><field2><type-id-2>...|

Reviewer: should `nfields` be `ncolumns`? and `field1` be `fieldname` ?

Author: Ok, I did `ncolumns` and `name1`...

`<ntypes>` cannot be 0.

#### 3.1.1.5 Enum Typedef

An enum type is encoded as the type code of the enum values as a `uvarint`
followed by a count of the number of elements in the enum, and in turn,
followed by the names and values of each element.
```
--------------------------------------------------------
|0xf4|<type-id>|<nelem>|<name1><val-1><name2><val-2>...|
--------------------------------------------------------
```
`<type-id>` and `<nelem>` are encoded as `uvarint`.

> The names have the same UTF-8 format as a record and are encoded

Reviewer: ```suggestion
The names have the same UTF-8 format as record field names and are encoded
```

as tag-encoded primitive strings.  Each value is encoded as
a tag-encoded value in accordance with the type indicated

> by ``<typd-id>`.

Reviewer: ```suggestion
by ``<type-id>`.
```

`<ntypes>` cannot be 0.

#### 3.1.1.5 Enum Typedef

An enum type is encoded as the type code of the enum values as a `uvarint`
followed by a count of the number of elements in the enum, and in turn,
followed by the names and values of each element.
```
--------------------------------------------------------
|0xf4|<type-id>|<nelem>|<name1><val-1><name2><val-2>...|
--------------------------------------------------------
```
`<type-id>` and `<nelem>` are encoded as `uvarint`.
The names have the same UTF-8 format as a record and are encoded
as tag-encoded primitive strings.  Each value is encoded as
a tag-encoded value in accordance with the type indicated
by ``<typd-id>`.

> Enums are typically integers but this representatiions, like the rust
> programming language, allows for arbitrarily-valued enumerations.

Reviewer: ```suggestion
Enums are typically integers but this representation, allows for arbitrarily-valued enumerations.
```

These enums are not like Rust enums

`<ntypes>` cannot be 0.

#### 3.1.1.5 Enum Typedef

An enum type is encoded as the type code of the enum values as a `uvarint`
followed by a count of the number of elements in the enum, and in turn,
followed by the names and values of each element.
```
--------------------------------------------------------
|0xf4|<type-id>|<nelem>|<name1><val-1><name2><val-2>...|
--------------------------------------------------------
```
`<type-id>` and `<nelem>` are encoded as `uvarint`.

> The names have the same UTF-8 format a record field names and are encoded

Reviewer 3: ```suggestion
The names have the same UTF-8 format as record field names and are encoded
```

as tag-encoded primitive strings.  Each value is encoded as
a tag-encoded value in accordance with the type indicated

> by ``<type-id>`.

Reviewer 3: ```suggestion
by `<type-id>`.
```

sequence must not include control messages.

> The reason control messages are not allowed in compressed blocks is to
> allow for optimizations that discard entire buffers of data based on
> heuristics to know a filtering predicate can't be true based on a
> quick scan of the data (e.g., using the Boyer-Moore algorithm to determine
> that a comparison with a string constant would not work for any record
> value in the buffer).  Since blocks may be dropped without parsing using
> such an optimization, any typedefs should be lifted out into the zng data
> stream in front of the compressed blocks (i.e., the stream is rearranged
> but it's always safe to move typedefs earlier in the stream as long as
> the typedef order is preserved and a zng end-of-stream is not crossed).
> For application-specific messages and end-of-stream, a compressed buffer
> should be terminated and these messages sent as uncompressed data.
>
> Since ZNG streams typically consist of a very sparse
> set of typedefs with very long runs of data, these constraints are not
> a barrier to performance in practice.

A compressed value message block is encoded as follows:
```

> ------------------------------------------------------------------------------
> |0xf6|<format><uncompressed-length>|<compressed-length>|<compressed-messages>|
> ------------------------------------------------------------------------------

Reviewer 2: ```suggestion
-------------------------------------------------------------------------------
|0xf6|<format>|<uncompressed-length>|<compressed-length>|<compressed-messages>|
-------------------------------------------------------------------------------
```
I botched this back when I added it.

It is also an error to redefine a previously defined alias with a
type that differs from the original definition.

### 3.1.2 Compressed Value Message Block

Following a header byte of `0xf6` is a compressed value message block.
Such a block comprises a compressed sequence of value messages.  The
sequence must not include control messages.

> The reason control messages are not allowed in compressed blocks is to
> allow for optimizations that discard entire buffers of data based on
> heuristics to know a filtering predicate can't be true based on a
> quick scan of the data (e.g., using the Boyer-Moore algorithm to determine
> that a comparison with a string constant would not work for any record
> value in the buffer).  Since blocks may be dropped without parsing using
> such an optimization, any typedefs should be lifted out into the zng data
> stream in front of the compressed blocks (i.e., the stream is rearranged

> // but its always safe to move typedefs earlier in the stream as long as

Reviewer 3: ```suggestion
// but it's always safe to move typedefs earlier in the stream as long as
```

Also, the `//` in there makes this paragraph look like it's still under construction. Not sure if it could be replaced with a simple comma or if this was some "XXX"-style placeholder to remind you to come back to it.

The header byte indicates the type ID of the value.  If the type ID
is larger than `0xde`, then the type ID is "escaped" with the value `0xdf`
and the actual type ID is encoded as a `uvarint` of the difference
of the type ID less the constant `0xdf`.

It is an error for a value to reference a type ID that has not been
previously defined by a typedef scoped to the stream in which the value
appears.

The value is encoded in the subsequent bytes using a "tag-encoding" scheme
that captures the structure of both primitive types and the recursive
nature of complex types.  This structure is encoded
explicitly in every value and the boundaries of each value and its
recursive nesting can be parsed without knowledge of the type or types of
the underlying values.  This admits an efficient implementation
for traversing the values, inclusive of recursive traversal of complex values,
whereby the inner loop need not consult and interpret the type ID of each element.

#### 3.2.1 Tag-Encoding of Values

Each value is prefixed with a "tag" that defines:
* whether it is a primitive or complex value,
* whether it is a nil value, and
* if non-nil, its encoded length in bytes.

> The list collection of sub-values comprising a value of a complex type

Reviewer: ```suggestion
The collection of sub-values comprising a value of a complex type
```

> is called a "container".

Reviewer: I'm not sure I get this definition... it sounds like it's saying that "container" is the contents ?

(Maybe "container" really refers to the encoding of a complex typed value?)

Author: I don't find it confusing so it would be helpful to get more clarity here.  I changed it to this: `The collection of sub-values comprising a complex-type value is called a "container".`  The low-level encoding represents a container of things, which is why I like the term for the spec and the code (not for user facing docs).  As a point of contrast, avro has complex values but they are not serialized with containers.  You decode the values in depth-first order and you need the complete type definition to make sense of the encoding.  With zng, containers let you decode the values without knowing the type.

Reviewer: The rewording is clear to me, thanks.

### 3.2 Value Messages

Following a header byte in the range `0x00-0xde` is a ZNG value.
The header byte indicates the type ID of the value.  If the type ID
is larger than `0xde`, then the type ID is "escaped" with the value `0xdf`
and the actual type ID is encoded as a `uvarint` of the difference
of the type ID less the constant `0xdf`.

It is an error for a value to reference a type ID that has not been
previously defined by a typedef scoped to the stream in which the value
appears.

The value is encoded in the subsequent bytes using a "tag-encoding" scheme
that captures the structure of both primitive types and the recursive
nature of complex types.  This structure is encoded
explicitly in every value and the boundaries of each value and its
recursive nesting can be parsed without knowledge of the type or types of
the underlying values.  This admits an efficient implementation
for traversing the values, inclusive of recursive traversal of complex values,
whereby the inner loop need not consult and interpret the type ID of each element.

#### 3.2.1 Tag-Encoding of Values

Each value is prefixed with a "tag" that defines:
* whether it is a primitive or complex value,

> * whether it is a nil value, and

Reviewer: I think some discussion of what `nil` means would be helpful. Likewise for `null`.

#### 3.2.2 Tag-Encoded Body of Primitive Values

Following the tag encoding is the value encoded in N bytes as described above.
A typed value with a `value` of length `N` is interpreted as described in the
[Primitive Types](#5-primitive-types) table.  The type information needed to
interpret all of the value elements of a complex type are all implied by the
top-level type ID of the value message.  For example, the type ID could indicate
a particular record type, which recursively provides the type information
for all of the elements within that record, including other complex types
embedded within the top-level record.

Note that because the tag indicates the length of the value, there is no need
to use varint encoding of integer values.  Instead, an integer value is encoded
using the full 8 bits of each byte in little-endian order.  For signed values,
before encoding, are shifted left one bit, and the sign bit stored as bit 0.
For negative numbers, the remaining bits are negated so that the upper bytes
tend to be zero-filled for small integers.

#### 3.2.2 Tag-Encoded Body of Complex Values

The body of a length-N container comprises zero or more tag-encoded values.

> Array, set, and record types are variable length and are encoded

Reviewer: ```suggestion
Array, set, and record values are variable length and are encoded
```

> as a sequence of elements and an enum as a small integer:

Reviewer: ```suggestion
as a sequence of elements and enum values as a small integer:
```

The above sentence doesn't mention `map` or `union`.  Rather than adding them, maybe drop the sentence as it's basically paraphrasing the table?

| Type     |          Value                          |
|----------|-----------------------------------------|
| `array`  | concatenation of elements               |
| `set`    | normalized concatenation of elements    |
| `record` | concatenation of elements               |

> | `union`  | concatenation of selector and value     |

Reviewer: ```suggestion
| `union`  | concatenation of type index and value     |
```

Author: This reads like a "type index" is a thing.  I like the term selector but how about "position"?

> | `enum`   | selector of id and value in enum type   |

Reviewer: ```suggestion
| `enum`   | index of name and value in enum type   |
```

Author: I used `position` as aboove and added further clarification in the text.

Reviewer: 👍

there is no need to encode a count of the
elements present.  Also, since the type ID is implied by the typedef
of any complex type, each value is encoded without its type ID.

For sets, the concatenation of elements must be normalized so that the
sequence of bytes encoding each element's tag-counted value is
lexicographically greater than that of the preceding element.

A union value is encoded as a container with two elements. The first
element is the `uvarint` encoding of the index determining the type of
the value in reference to the union type, and the second element is
the value encoded according to that type.

An enumeration value is simply a single, small integer encoded as length-N
integer that represents the positional index of enum identifier and value as it
is defined in the enum typedef.

A map value is encoded as a container as a sequence of alternating tag-encoded
key and value. The concatenation of elements must be normalized so that the
sequence of bytes encoding each tag-counted key (of the key/value pair) is
lexicographically greater than that of the preceding key (of the preceding
key/value pair).

> > XXX it's not clear that we should require sorted keys and give up

Reviewer: I'm not understanding this sentence... is the deletion of the paragraph describing set normalization suggesting that we remove that requirement ? (I don't see how we can, hence part of my confusion here).

Author: This was moved to line 573.  The XXX refers to map values.  Maybe we can discuss at sync-up?

Reviewer: Ah hadn't spotted that the paragraph moved. This makes sense.

`<type-name>` is an identifier with semantics as defined in [Section 3.1.1.5](#3115-alias-typedef).

### 4.1.3 Application-Defined Messages

A TZNG application-defined message has the following form:
```
#!<control-code>:<message>
```
Here, `<control-code>` is a decimal integer in the range 248-254 and `<message>`
is any UTF-8 string with escaped newlines.

### 4.1.4 End of Stream

> A TZNG end-of-stream market has the following form:

Reviewer: ```suggestion
A TZNG end-of-stream marker has the following form:
```

Author: The market for end-of-streams is huge!

```
#!255
```
A TZNG stream or file should always be terminated with end-of-stream.
This clears all of the previous type tag bindings and aliases, allowing
multiple TZNG files with overlapping tags to be concatenated without the
tags colliding.

> TBD: implement TZNG EOS in zq

### 4.2 Type Grammar

Given the above textual definitions and the underlying ZNG specification, a
grammar describing the textual type encodings is:
```

> <stype> := bool | init8 | uinit8 | int16 | uint16 | int32 | uint32 | int64 | uint64 | float64

Reviewer: ```suggestion
<stype> := bool | int8 | uint8 | int16 | uint16 | int32 | uint32 | int64 | uint64 | float64
```

`<type-name>` is an identifier with semantics as defined in [Section 3.1.1.5](#3115-alias-typedef).

### 4.1.3 Application-Defined Messages

A TZNG application-defined message has the following form:
```
#!<control-code>:<message>
```
Here, `<control-code>` is a decimal integer in the range 248-254 and `<message>`
is any UTF-8 string with escaped newlines.

### 4.1.4 End of Stream

A TZNG end-of-stream marker has the following form:
```
#!255
```
A TZNG stream or file should always be terminated with end-of-stream.
This clears all of the previous type tag bindings and aliases, allowing
multiple TZNG files with overlapping tags to be concatenated without the
tags colliding.

> > TBD: implement TZNG EOS in zq

Reviewer 3: Should this just be moved to the top of the doc with the other zq "not implemented" exceptions? Or is it here because you intend to implement it really soon?

Author: Yes, also am changing this to `#eos`.

[ ]
```
In addition, `-` must be escaped if representing the single ASCII byte equal

> to `-` as opposed to representing an null value.

Reviewer: ```suggestion
to `-` as opposed to representing a null value.
```

| `bytes`    | 14 | variable | N bytes of value                               | a sequence of bytes encoded as base64                         |
| `string`   | 15 | variable | UTF-8 byte sequence of string                  | a UTF-8 string                                                |
| `bstring`  | 16 | variable | UTF-8 byte sequence with `\x` escapes          | a UTF-8 string with `\x` escapes of non-UTF binary data       |
| `ip`       | 17 | 4 or 16  | 4 or 16 bytes of IP address                    | a string representing an IP address in [IPv4 or IPv6 format](https://tools.ietf.org/html/draft-main-ipaddr-text-rep-02#section-3) |
| `net`      | 18 | 8 or 32  | 8 or 32 bytes of IP prefix and subnet mask     | a string in CIDR notation representing an IP address and prefix length as defined in RFC 4632 and RFC 4291. |

> | `error`    | 19 | variable | UTF-8 byte sequence of string of error message | a string in CIDR notation representing an IP address and prefix length as defined in RFC 4632 and RFC 4291. |

Reviewer: ```suggestion
| `error`    | 19 | variable | UTF-8 byte sequence of string of error message | a UTF-8 string. |
```

> | `type`     | 20 | variable | unsigned int of length N representing type ID  | a string in CIDR notation representing an IP address and prefix length as defined in RFC 4632 and RFC 4291. |

Reviewer: ```suggestion
| `type`     | 20 | variable | UTF-8 byte sequence of string representing type in the [TZNG type grammar](#42-type-grammar)   | a UTF-8 string  |
```

(right?)

Author: yes

#26:record[city:string,lat:LL,long:LL]
26:[NYC;[N;40.7128;][W;74.0060;]]
```

> An null value indicates a field of a `record` that wasn't set by the encoder:

Reviewer: ```suggestion
A null value indicates a field of a `record` that wasn't set by the encoder:
```

Author: Ok, adding more on null value and type null.  Getting rid of "nil".  We did that bikeshed :-)
